### ISSUE TITLE
Conditional validation rules

### ISSUE THREAD

**Issue by stijnvanhoey** [nicolaiarocci/cerberus#206]
When I'm checking the issues, the idea of a conditional sets of tests does appear regularly in different kinds of similar requests, eg. #121, #136, #182. 

Part of it can be resolved by the *rules and part is covered by the `dependencies` rule. However, as I understand it, it would be beneficial if there would be a generalization of the dependencies, currently checking the rule  'if present', towards any set of conditional rules (set of validation rules). As mentioned by @fredrikwidlund, putting them all in oneof-statements gets really extensive. It comes down to some kind of general conditional environment, cfr. cfr. comment [comment](https://github.com/nicolaiarocci/cerberus/issues/136#issuecomment-139277299) of issue: #136

The idea is that there would be a conditional set of rules, providing a number of tests that, when valid, invokes a set of tests for the specified field. 
Consider the following example. We have a field with the perimeter and a field with the size description: The rule: `medium` is only allowed for the field `size`, when the field `perimeter` is in between `10` and `30`. Using a conditional schema, for example with the introduction of an if statement: 

```python
schema = {'size': {'if': {'perimeter': {'min': 10, 'max': 30}, 'allowed': 'medium'}}}
```
Within the if, all field names would induce a set of conditional rules (e.g. `min`/`max` on `perimeter`), all others are rules that need to be tested on field size (e.g. allowed of `medium`) when conditions provided by other fields do comply. This would mean that the following document would be ok:
```python
document = {'size' : 'medium', 'perimeter': 20}
```
But this one would provide an error on the size
```python
document = {'size' : 'large', 'perimeter': 20}
```

I drafted a custom if-validator according to the documentation (but using an instance creation within the definition which is maybe not ideal...):
```python
class MyValidator(Validator):
    def _validate_if(self, ifset, field, value):
        # extract dict values -> conditions
        conditions = {k: v for k, v in ifset.iteritems() if isinstance(v, dict)}
        # extract dict values -> rules
        rules = {k: v for k, v in ifset.iteritems() if not isinstance(v, dict)}

        valid=True
        # check for all conditions if they apply
        for term, cond in conditions.iteritems():
            subschema = {term : cond}
            tempvalidation = MyValidator(subschema)
            tempvalidation.allow_unknown = True
            if not tempvalidation.validate(self.document):
                valid=False

        #others -> conditional rules applied when valid condition
        if valid:
            tempvalidation = MyValidator({field: rules})
            tempvalidation.validate({field : self.document[field]})
            #convert to object itself
            for field, err in tempvalidation.errors.items():
                self._error(field, err)
```

using this code snippet:
```python
val = MyValidator(schema)
val.allow_unknown = True
val.validate(document)
```
provides indeed an error on the `{'size': 'unallowed value large'}` when using `document = {'size' : 'large', 'perimeter': 20}`...
The idea is that the conditions could be any combination of rules (min, max, maxlength, empty,...) invoking tests for the filed when conditions are valid.

I realize that the `if` workaround my not be the optimal decision, some kind of mapping could be more interesting, maybe an if/elif/.../else version?  I would like to see if there would be the possibility of bringing together the issues related to the *conditional* field validators in an clear and transparent manner. 

So, I would like to ask the following:
* taking into account that a large set of these rules would be required, the `oneof` with multiple version of a schema would become very extensive; is there another way (combination of existing rules,...) currently provided that would support this kind of conditional testing? 
* If not, is the provided snippet with custom validator a first way to go or are there alternative options to set up such a conditional testing-environment?

**Comment by funkyfuture** [comment]
schema = {'size': {'if': {'perimeter': {'min': 10, 'max': 30}, 'allowed': 'medium'}}}

> Within the if, all field names would induce a set of conditional rules (e.g. min/max on perimeter), all others are rules that need to be tested on field size (e.g. allowed of medium) when conditions provided by other fields do comply. 

in that proposal it isn't clear what the condition and what the conditional rule is. it's not posssible to figure that out as Python dictionaries are not sorted. (in your reference implementation you're missing that rules can also have mappings as constraints.) so, this would rather work and be explicit:

    schema = {<field>: {'if': {<field>: <constraints>, ...}, 'then': {<constraints>, 'else': <constraints>}

i wouldn't take an `elif` into the grammar as there could only be one and that would lead to confusion / not be pythonic.

but still, your example is a perfect case for an `oneof`. the functionality is covered. and it allows more than two conditionals. (how would `small` and `extralarge` fit in your example?)

i see that in yaml your proposal is neat:

```yaml
field:
  if: ...
  then: ...
  else: ...
```

but, well, that could simply be translated into an `oneof` schema.

my impression so far is, that these requests are more a documentation issue. maybe it would help to show various examples with the *of-rules.

**Comment by nicolaiarocci** [comment]
Agree with @funkyfuture although I agree that an "if" rule would surely more easy to grasp for everyone. Updating the docs with more use cases/examples is the best possible course imho.

**Comment by funkyfuture** [comment]
@annakz  maybe [this question and answers](https://stackoverflow.com/q/42482923/2489914) can help you. otherwise i propose you ask a question with a concrete use-case there too.

**Comment by jeking3** [comment]
I have had the need for "conditional required" logic as well, but I see the authors rejected the proposal.  It's a shame, because the prescribed way to implement conditional logic like that through "*of" branches.  The problem I have with using *of branches are twofold:
1. It violates the principal of eliminating duplication.  When you have many custom rules to carry metadata through to processing (for example, a description), you end up with a lot of duplication.  Duplication is something folks try to avoid when programming, but *of rules to implement conditional constraints actually encourages duplication of validation information in this case.
2. When using code generation on the schema yaml to build a python class to make it easier to access things like default values (also a different custom rule I made), interpreting "*of" to understand what the properties should be for the class is very difficult.

As such I've implemented a flexible "constraint" rule and a subclass Validator to implement it, using objectpath as the query language.  The schema for this is:

```schema validation schema
      constraints:
        # Enforces constraints by evaluating an objectpath statement.
        # The statement has access to all of the data at the current
        # level of the document.  The logic is as follows:
        # 1. Test the "if" statement.  If True, then..
        # 2. Test the "then" statement.  If True, constraint is satisfied.
        #    If False, then...
        # 3. Mark the property indicated by "property" with a message in
        #    "else".
        type: list
        required: false
        empty: false
        schema:
          type: dict
          required: true
          schema:
            if:
              # objectpath expression must evaluate to True to proceed to...
              type: string
              required: true
              empty: false
            then:
              # objectpath expression must evanuate to True or raise...
              type: string
              required: true
              empty: false
            else:
              # error message to apply to...
              type: string
              required: true
              empty: false
            property:
              # property to apply the error message to
              type: string
              required: true
              empty: false
```

The implementation is rather simple - inside a Validator subclass:
```python
    def _enforce_one(self, constraint: dict, data):
        """
        Enforce a single constraint against data.

        The constraint dictionary consists of:
        - if: This objectpath query must evaluate to True to engage enforcement
        - then: This objectpath query must evaluate to True to pass enforcement
        - else: This is the message applies to a property if enforcement fails
        - property: The name of the property to apply the message to
        """
        tree = objectpath.Tree(data)
        rqif = tree.execute(constraint["if"])
        self._logger.debug(f"enforce if {constraint['if']}: {rqif}")
        if rqif:
            # We met the "if" criteria.  Now test "then"
            rqthen = tree.execute(constraint["then"])
            self._logger.debug(f"enforce then {constraint['then']}: {rqthen}")
            if not rqthen:
                # We met the "if" criteria but failed the "then" criteria
                # This means enforcement failed
                self._error(constraint["property"], constraint["else"])

    def _validate_constraints(self, instance, field, value):
        """
        Allows for constraints to be applied that are complex or
        difficult to achieve with other tools like cerberus, such
        as conditionally required fields.  To guarantee this rule
        executes in all cases it must exist on a required field.

        The rule's arguments are validated against this schema:
        {
          'type': 'list',
          'required': False,
          'empty': False,
          'schema': {
            'type': 'dict',
            'required': True,
            'schema': {
              'if': {
                'type': 'string',
                'required': True,
                'empty': False
              },
              'then': {
                'type': 'string',
                'required': True,
                'empty': False
              },
              'else': {
                'type': 'string',
                'required': True,
                'empty': False
              },
              'property': {
                'type': 'string',
                'required': True,
                'empty': False
              }
            }
          }
        }
        """
        for constraint in instance:
            self._enforce_one(constraint, value)
```

Concrete example of using this is:
```
        schema:
          private:
            description: |-
                IP address information for the IP address visible to the attached
                device.
            type: dict
            required: true
            constraints:
              # cerberus does not have "conditionally required" logic,
              # so we made our own
              - if: $.type is 'static'
                then: $.ip
                else: required when type is static
                property: ip
            schema:
              ip:
                description: |-
                    A private IP address assigned to the {{type}}.  This field is required
                    if `type` is `static`.
                type: string
                required: false
                # required-if: $.parent.type is 'static'
                content: ip
                # announce: true
              type:
                description: |-
                    The allocation type of the private IP address.  If this field is `static`
                    then the `ip` field is required.
                type: string
                required: true
                allowed:
                  - dynamic
                  - static
```

The only gotcha I see with this is that you have to apply it to a required field so that the rule is always processed, as if you put it on an optional field, and the field is not present, then the rule doesn't fire.

**Comment by funkyfuture** [comment]
how is this not duplicating and what would it mean to a human consumer?

```yaml
            if:
              # objectpath expression must evaluate to True to proceed to...
              type: string
              required: true
              empty: false
            then:
              # objectpath expression must evanuate to True or raise...
              type: string
              required: true
              empty: false
            else:
              # error message to apply to...
              type: string
              required: true
              empty: false
            property:
              # property to apply the error message to
              type: string
              required: true
              empty: false
```

**Comment by jeking3** [comment]
The alternative to the schema snippet I proposed using "oneof" rules would contain duplication of the schema to address both cases, which is a fairly simple schema.  In a more complex schema with multiple conditional required constraints it would be even worse, one duplication of a part of the schema for each of the possible constraints.  Even with that, code generation for classes derived from yaml schema using the "*of" branches are more difficult to process.   I'll put together a schema that implements the conditional with "oneof" logic for comparison.

Basically what I was looking for was the "required-if" rule that can be found commented out in my other post.
